packages/dd-trace/src/llmobs/plugins/genai/util.js

```diff
@@ -0,0 +1,499 @@
+
+  for (const candidate of response.candidates) {
+    const content = Array.isArray(candidate) ? candidate[0].content : candidate.content
+    if (!content?.parts) continue
+
+    // Skip special cases in streaming (handle them as non-streaming)
+    if (content.parts.some(part => part.functionCall ||
+                                   part.executableCode ||
+                                   part.codeExecutionResult)) {
+      messages.push(...formatNonStreamingCandidate(candidate))
+      continue
+    }
+
+    // Accumulate text parts by role
+    const partsByRole = groupPartsByRole(content.parts)
+
+    for (const [partRole, textContent] of Object.entries(partsByRole)) {
+      if (!textContent) continue
+
+      if (messagesByRole.has(partRole)) {
+        const index = messagesByRole.get(partRole)
+        messages[index].content += textContent
+      } else {
+        const messageIndex = messages.length
+        messages.push({ role: partRole, content: textContent })
+        messagesByRole.set(partRole, messageIndex)
+      }
+    }
+  }
+
+  return messages.length > 0 ? messages : [{ content: '' }]
+}
+
+/**
+ * Format non-streaming output from response
+ * @param {object} response
+ * @returns {Array}
+ */
+function formatNonStreamingOutput (response) {
+  const messages = []
+
+  for (const candidate of response.candidates) {
+    messages.push(...formatNonStreamingCandidate(candidate))
+  }
+
+  return messages.length > 0 ? messages : [{ content: '' }]
+}
+
+/**
+ * Format output messages from response
+ * @param {object} response
+ * @param {boolean} isStreaming
+ * @returns {Array}
+ */
+function formatOutputMessages (response, isStreaming = false) {
+  if (!response?.candidates?.length) {
+    return [{ content: '' }]
+  }
+
+  if (isStreaming) {
+    return formatStreamingOutput(response)
+  }
+
+  return formatNonStreamingOutput(response)
+}
+
+/**
+ * Format embedding output from response
+ * @param {object} response
+ * @returns {string}
+ */
+function formatEmbeddingOutput (response) {
+  if (!response?.embeddings?.length) {
+    return ''
+  }
+
+  const embeddingCount = response.embeddings.length
+  const firstEmbedding = response.embeddings[0]
+
+  if (firstEmbedding.values && Array.isArray(firstEmbedding.values)) {
+    const embeddingDim = firstEmbedding.values.length
+    return `[${embeddingCount} embedding(s) returned with size ${embeddingDim}]`
+  }
+
+  return `[${embeddingCount} embedding(s) returned]`
+}
+
+module.exports = {
+  ROLES,
+  getOperation,
+  extractTextParts,
+  groupPartsByRole,
+  hasThoughtParts,
+  determineRole,
+  normalizeRole,
+  extractMetrics,
+  extractMetadata,
+  formatFunctionCallMessage,
+  formatFunctionResponseMessage,
+  aggregateStreamingChunks,
+  formatContentObject,
+  formatInputMessages,
+  formatEmbeddingInput,
+  formatNonStreamingCandidate,
+  formatStreamingOutput,
+  formatNonStreamingOutput,
+  formatOutputMessages,
+  formatEmbeddingOutput
+}
```

Comment:
from the `index.js` file, it looks like we only ever use

```
getOperation
extractMetrics
extractMetadata
aggregateStreamingChunks
formatInputMessages
formatEmbeddingInput
formatOutputMessages
formatEmbeddingOutput
```

so we can probably trim down what we export from this file to just that.

otherwise the contents of this file lgtm!